CommuniTEAM/CommuniTEA
# [TEA-72] Users Endpoints (part 2)

### In this PR, I'm adding/changing the following:

#### Additions

- Added new endpoints for Users to achieve a full CRUD
  - GetUser - GET endpoint to grab details of any user
  - UpdateUser - PUT endpoint to update details of your own user account
  - ChangePassword - PUT endpoint to change your account password
  - DeleteUser - DELETE endpoint to delete your own user account
- Added users SQL queries to support the endpoints
- Made a helper functions to reduce cognitive complexity of the beefier endpoints (UpdateUser... :smiling_face_with_tear:)
  - `verifyEmail` takes an email as a string, a database query connection, and a user UUID as a string. It returns a wrapped error if the email does not match email regex or if the email is already in the database for a user with a different UUID
- Added a test suite for Users endpoints in `users_test.go`

#### Fixes / Improvements

- Fixed the backend deployment workflow: the `upload-artifact` action wasn't using the specified working directory
- Improved the data shape of user outputs so that the location includes the location ID, city, and state
- Fixed the return data type of `ValidateJWT()` in `tokens.go` to return `TokenData` instead of a map interface
- Fixed the `TokenData` struct to include the user's email
- Lowered target code coverage from 90% to 80% because anything that returns a 500 error counts as uncovered, which lowers the coverage scores considerably.
- Made the linter less strict with function max line numbers (increased to 125) and max cognitive complexity (increased to 35)

### Please do the following to test my code:

1. Delete and rebuild docker containers
2. Test all of the Users endpoints at localhost:8000/docs
3. Check [the code coverage report](https://app.codecov.io/gh/CommuniTEAM/CommuniTEA/pull/44/blob/backend/api/users.go) for Users to ensure that all lines for endpoints **other than UpdateUsers** (aside from 500 errors) are covered.

### Notes

The tests for UpdateUser are not complete! I will make a separate PR with tests for `auth.go` and flesh out the tests for this endpoint in that, as well.

[Related Jira Ticket](https://communiteam.atlassian.net/browse/TEA-72)
